src/main/java/org/radarcns/stream/collector/NumericAggregateCollector.java

```diff
@@ -0,0 +1,203 @@
+ */
+
+package org.radarcns.stream.collector;
+
+import static org.radarcns.util.Serialization.floatToDouble;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Field;
+import org.apache.avro.Schema.Type;
+import org.apache.avro.specific.SpecificRecord;
+
+/**
+ * Java class to aggregate data using Kafka Streams. Double is the base type.
+ * Only the sum and sorted history are collected, other values are calculated on request.
+ */
+public class NumericAggregateCollector implements RecordCollector {
+    private final String name;
+    private final int pos;
+    private final Type fieldType;
+    private BigDecimal sum;
+    private final List<Double> history;
+
+    @JsonCreator
+    public NumericAggregateCollector(
+            @JsonProperty("name") String name, @JsonProperty("pos") int pos,
+            @JsonProperty("fieldType") Type fieldType, @JsonProperty("sum") BigDecimal sum,
+            @JsonProperty("history") List<Double> history) {
+        this.name = name;
+        this.pos = pos;
+        this.fieldType = fieldType;
+        this.sum = sum;
+        this.history = new ArrayList<>(history);
+    }
+
+    public NumericAggregateCollector(String fieldName) {
+        this(fieldName, null);
+    }
+
+    public NumericAggregateCollector(String fieldName, Schema schema) {
+        sum = BigDecimal.ZERO;
+        this.history = new ArrayList<>();
+
+        this.name = fieldName;
+        if (schema == null) {
+            this.pos = -1;
+            this.fieldType = null;
+        } else {
+            Field field = schema.getField(fieldName);
+            if (field == null) {
+                throw new IllegalArgumentException(
+                        "Field " + fieldName + " does not exist in schema " + schema.getFullName());
+            }
+            this.pos = field.pos();
+
+            Type apparentType = field.schema().getType();
+            if (apparentType == Type.UNION) {
+                for (Schema subSchema : field.schema().getTypes()) {
+                    if (subSchema.getType() != Type.NULL) {
+                        apparentType = subSchema.getType();
+                        break;
+                    }
+                }
+            }
+            fieldType = apparentType;
+
+            if (fieldType != Type.DOUBLE
+                    && fieldType != Type.FLOAT
+                    && fieldType != Type.INT
+                    && fieldType != Type.LONG) {
+                throw new IllegalArgumentException("Field " + fieldName + " is not a number type.");
+            }
+        }
+    }
+
+    @Override
+    public NumericAggregateCollector add(SpecificRecord record) {
+        if (pos == -1) {
+            throw new IllegalStateException(
```

Comment:
In this case, should we have the `public NumericAggregateCollector(String fieldName)` constructor? I do not see another way of setting a value for `pos`.

Reply:
In that case it is still possible to create derived aggregates. So for Empatica E4, we can derive heart rate from inter-beat-interval, by just adding a value as a double and giving it the "heartRate" field name. However, then the `add(SpecificRecord)` method may not be used.